tests/app/questionnaire/test_value_source_resolver.py

```diff
@@ -0,0 +1,371 @@
+            [{"answer_id": "some-answer", "list_item_id": "item-1", "value": "Yes"}]
+        ),
+        list_item_id="item-1",
+    )
+
+    assert value_source_resolver.resolve(answer_source) is None
+
+    value_source_resolver = get_value_source_resolver(
+        schema=schema,
+        answer_store=AnswerStore([{"answer_id": "some-answer", "value": "Yes"}]),
+        list_item_id="item-1",
+    )
+
+    assert value_source_resolver.resolve(answer_source) == "Yes"
+
+
+def test_answer_source_with_list_item_selector_location():
+    value_source_resolver = get_value_source_resolver(
+        answer_store=AnswerStore(
+            [
+                {
+                    "answer_id": "some-answer",
+                    "list_item_id": "item-1",
+                    "value": "Yes",
+                }
+            ]
+        ),
+        location=Location(
+            section_id="some-section", block_id="some-block", list_item_id="item-1"
+        ),
+    )
+
+    assert (
+        value_source_resolver.resolve(answer_source_list_item_selector_location)
+        == "Yes"
+    )
+
+
+def test_answer_source_with_list_item_selector_list_first_item():
+    value_source_resolver = get_value_source_resolver(
+        answer_store=AnswerStore(
+            [
+                {
+                    "answer_id": "some-answer",
+                    "list_item_id": "item-1",
+                    "value": "Yes",
+                }
+            ]
+        ),
+        list_store=ListStore([{"name": "some-list", "items": get_list_items(3)}]),
+    )
+
+    assert (
+        value_source_resolver.resolve(answer_source_list_item_selector_list_first_item)
+        == "Yes"
+    )
+
+
+@pytest.mark.parametrize("is_answer_on_path", [True, False])
+@pytest.mark.parametrize("is_inside_repeat", [True, False])
+def test_answer_source_with_routing_path_block_ids(is_answer_on_path, is_inside_repeat):
+    schema = get_mock_schema()
+    schema.get_block_for_answer_id = Mock(return_value={"id": f"some-block"})
+
+    location = Location(section_id="test-section", block_id="test-block")
+    id_prefix = "some" if is_answer_on_path else "some-other"
+    answer = Answer(answer_id=f"{id_prefix}-answer", value="Yes")
+
+    if is_inside_repeat:
+        location.list_item_id = answer.list_item_id = "item-1"
+        schema.answer_should_have_list_item_id = Mock(return_value=True)
+    else:
+        schema.answer_should_have_list_item_id = Mock(return_value=True)
+
+    value_source_resolver = get_value_source_resolver(
+        schema=schema,
+        answer_store=AnswerStore([answer.to_dict()]),
+        list_store=ListStore([{"name": "some-list", "items": get_list_items(3)}]),
+        location=location,
+        list_item_id=location.list_item_id,
+        routing_path_block_ids=[f"{id_prefix}-block"],
+    )
+
+    expected_result = "Yes" if is_answer_on_path else None
+    assert value_source_resolver.resolve(answer_source) == expected_result
```

Comment:
> Looking into this further there is an _is_answer_on_path method in value_source_resolver.py and also in rules.py. I think it would make sense to have a common public method method that they both use. This could make this test clearer as we could mock the is_answer_on_path method instead.

I purposely duplicated it as I think it should live in value source resolver and not a public method. There is no other use case apart from needing it for this, and I didn't want to mix the old routings methods with this. My intentions were that once we support the new routing rules fully, the old code can be dropped.

> Naming the ids a bit more clearly could help to show the intent of the test e.g. block-on-path rather than some-block etc.

Have done this to try to make it clearer.

> It's confusing that schema.get_block_for_answer_id is mocked to return the same id regardless of answer, and too subtle that this is then used to match or not match the value for routing_path_block_ids.

As my point above, if the changes I made have not made it clearer or clear enough, I rather split this test in two or make `_is_answer_on_path` public to ValueSourceResolver and then patch that. Let me know what you think.

Reply:
Fair point on not mixing up this work with routing code that is being replaced. However, the old routing code checks that the answers used for routing are on the path. I assume this is needed for the new routing changes as well, so it is something we should consider factoring out (possibly when the routing work is being done).

The new naming does make the test clearer, but I still think the fixed return value for `schema.get_block_for_answer_id` is confusing. Flipping it so that method returns an appropriate block id and passing a fixed block id as the `routing_path_block_ids` param to `get_value_source_resolver` makes it much clearer.

```python
    if is_answer_on_path:
        answer_id = "answer-on-path"
        schema.get_block_for_answer_id = Mock(return_value={"id": f"block-on-path"})
    else:
        answer_id = "answer-not-on-path"
        schema.get_block_for_answer_id = Mock(return_value={"id": f"block-not-on-path"})
```
and
```python
value_source_resolver = get_value_source_resolver(
        schema=schema,
        answer_store=AnswerStore([answer.to_dict()]),
        list_store=ListStore([{"name": "some-list", "items": get_list_items(3)}]),
        location=location,
        list_item_id=location.list_item_id,
        routing_path_block_ids=["block-on-path"],
    )
```

I think it would be clearer if the test was split into two, so that repeating is done separately.

Reply:
> However, the old routing code checks that the answers used for routing are on the path. I assume this is needed for the new routing changes as well, so it is something we should consider factoring out (possibly when the routing work is being done).

As discussed, the new routing work uses `ValueSourceResolver` hence only ValueSources need access to `is_answer_on_path`.

> The new naming does make the test clearer, but I still think the fixed return value for schema.get_block_for_answer_id is confusing. Flipping it so that method returns an appropriate block id and passing a fixed block id as the routing_path_block_ids param to get_value_source_resolver makes it much clearer.

Made these changes and have also split the test into two and accepted some duplication to improve readability.